src/arch/wasm/CodeGen.zig

```diff
@@ -5821,6 +5821,152 @@ fn airPopcount(func: *CodeGen, inst: Air.Inst.Index) InnerError!void {
+
+            if (bits == 128) {
+                try func.store(.stack, second_half_reversed, Type.u64, result.offset() + 8);
+                try func.store(.stack, first_half_reversed, Type.u64, result.offset());
+            } else {
+                var tmp = try func.allocLocal(Type.u64);
+                defer tmp.free(func);
+                try func.addLabel(.local_tee, tmp.local.value);
+                try func.emitWValue(.{ .imm64 = 128 - bits });
+                try func.addTag(.i64_shr_u);
+                try func.store(.stack, .stack, Type.u64, result.offset() + 8);
+                try func.addLabel(.local_get, tmp.local.value);
+                try func.emitWValue(.{ .imm64 = bits - 64 });
+                try func.addTag(.i64_shl);
+                try func.addTag(.i64_or);
+                try func.store(.stack, .stack, Type.u64, result.offset());
+            }
+            func.finishAir(inst, result, &.{ty_op.operand});
+        },
+        else => unreachable,
+    }
+}
+
+/// Reverses the integer that stored in wasm i32 value, with specified size in bits
+/// NOTE: Leaves the result on the stack.
+fn bitReverseU32(func: *CodeGen, value: WValue, bits: u16) InnerError!WValue {
+    var tmp = try func.allocLocal(Type.u32);
+    defer tmp.free(func);
+    try func.emitWValue(value);
+    for ([_]struct { mask: u32, stride: u32 }{
+        .{ .mask = 0x55555555, .stride = 1 },
+        .{ .mask = 0x33333333, .stride = 2 },
+        .{ .mask = 0x0F0F0F0F, .stride = 4 },
+        .{ .mask = 0x00FF00FF, .stride = 8 },
+        .{ .mask = 0x0000FFFF, .stride = 16 },
+    }) |step| {
+        if (2 * step.stride >= bits) {
+            try func.addLabel(.local_tee, tmp.local.value);
+            try func.emitWValue(.{ .imm32 = 2 * step.stride - (bits - step.stride) });
+            try func.addTag(.i32_shr_u);
+            try func.addLabel(.local_get, tmp.local.value);
+            try func.emitWValue(.{ .imm32 = step.mask });
+            try func.addTag(.i32_and);
+            try func.emitWValue(.{ .imm32 = bits - step.stride });
+            try func.addTag(.i32_shl);
+            try func.addTag(.i32_or);
+            break;
+        }
+
+        try func.addLabel(.local_tee, tmp.local.value);
+        try func.emitWValue(.{ .imm32 = step.stride });
+        try func.addTag(.i32_shr_u);
+        try func.emitWValue(.{ .imm32 = step.mask });
+        try func.addTag(.i32_and);
+        try func.addLabel(.local_get, tmp.local.value);
+        try func.emitWValue(.{ .imm32 = step.mask });
+        try func.addTag(.i32_and);
+        try func.emitWValue(.{ .imm32 = step.stride });
+        try func.addTag(.i32_shl);
+        try func.addTag(.i32_or);
+    }
+    return .stack;
+}
+
+/// Reverses the integer that stored in wasm i64 value, with specified size in bits
+/// NOTE: Leaves the result on the stack.
+fn bitReverseU64(func: *CodeGen, value: WValue, bits: u16) InnerError!WValue {
```

Comment:
From my perspective it seems like this algorithm belongs in compiler-rt. If implemented in portable zig code in compiler-rt it would be useful for all targets that don't have a native bit reverse instruction.

Have you considered that possibility?

Reply:
No, I did not, but I agree with your thoughts. There is an advantage of doing it directly in the backend because it leads to executing significantly less instructions for integers smaller than 32 bits, still the portable zig code + help other backends (probably all) is outweighting it.

I am thinking of doing it, but I never touched compiler-rt and do not know its naming conventions. @Luukdegram do you have any remarks about this?

Reply:
I just found `bswap.zig` in compiler-rt is similar in nature, but operates on byte granularity, so this should be easy to copy.